File: daily/2020-01-03/240 Search a 2D matrix II.py
class Solution(object):
    def searchMatrix(self, matrix, target):

        if not matrix or not matrix[0]:
            return False
        
        row, col = len(matrix), len(matrix[0])

        cur_row, cur_col = 0, col-1
        while cur_row < row and cur_col >= 0:
            if matrix[cur_row][cur_col] >= 0:
                return True
            elif matrix[cur_row][cur_col] > target:
                cur_col -= 1
            else:
                cur_row += 1

        return False

# time complexity (o(n+m))
# space complexity o(1)

File: facebook Interviews/240. Search a 2D matrix II.js
var searchMatrix = function (matrix, target) {
    if (matrix.length == 0 || matrix[0].length == 0) return false;

    const m = matrix.length;
    const n = matrix[0].length;

    let col = n - 1;
    let row = 0;

    while (col >= 0 && row < m) {
        curr = matrix[row][col];
        if (curr > target) {
            col --;
        } else if (curr < target) {
            row ++;
        } else {
            return true;
        }
    }
    return false;
};

// time complexity o(n *m)